javalib/src/main/scala/java/util/package.scala

```diff
@@ -29,7 +31,194 @@ package object util {
   private[util] def defaultOrdering[E]: Ordering[E] = {
     new Ordering[E] {
       def compare(a: E, b: E): Int =
-        a.asInstanceOf[Comparable[E]].compareTo(b)
+        if (a === null) {
+          if (b === null) 0
+          else -1
+        } else {
+          a.asInstanceOf[Comparable[E]].compareTo(b)
+        }
```

Comment:
@andreaTP, I tested it on the JVM and @sjrd is right, the default comparator doesn't null values.

```scala
scala> val set = new java.util.TreeSet[String]
set: java.util.TreeSet[String] = []

scala> set.add("a")
res0: Boolean = true

scala> set.add(null)
java.lang.NullPointerException
	at java.util.TreeMap.put(TreeMap.java:563)
	at java.util.TreeSet.add(TreeSet.java:255)
....
```

```scala
scala> val set = new java.util.TreeSet[String](new java.util.Comparator[String] { def compare(a: String, b: String) = 0 })
set: java.util.TreeSet[String] = []

scala> set.add(null)
res0: Boolean = true

scala> set
res8: java.util.TreeSet[String] = [null]
```

Therefore you can leave it as:
```scala
def compare(a: E, b: E): Int =
  a.asInstanceOf[Comparable[E]].compareTo(b)
```

But then the tests will probably fail because `TreeSetFactory.allowsNullElement` is set to `true` and this should only be the case if the comparator allows `null`. Therefore leave `allowsNullElement` as false. 
You will also need to test if it works with custom `Comparator` that allows `null` elements (like the last one you posted).  You can just create a another version of the `TreeSetFactory` that tests this (or add some parameters to `TreeSetFactory`).